docs/contracts/v4/concepts/05-lock-mechanism.mdx

```diff
@@ -0,0 +1,212 @@
+### 4. **Restricted Access:**
+
+  - The modifier `onlyByLocker` is used to restrict access to certain functions. It ensures that a function can only be
+called by the address that currently holds the lock.
+  ```solidity
+    modifier onlyByLocker() {
+        address locker = lockData.getActiveLock();
+        if (msg.sender != locker) revert LockedBy(locker);
+        _;
+    }
+  ```
+  - Several functions in the contract, such as `modifyPosition`, `swap`, `donate`, `take`, `mint`, and `settle`, use
+the `onlyByLocker` modifier. This ensures that only the entity that has currently locked the contract can modify
+positions, swap tokens, etc. This helps ensure operations are atomic and avoid potential race conditions.
+
+# Locking Diagram
+
+![Locking Diagram](./images/05_locking_mechanism/LockingMechanism_excali.png)
+
+* A locker (caller) initiates a lock function execution in the PoolManager contract.
+* The locker gets added to the lock queue by the LockDataLibrary.
+* An ILockCallback is used where the locker acquires the lock.
+* Balances and currency deltas are managed and updated within the Balance and CurrencyDelta.
+* The locker can account for pool balance deltas which handle the currency and its corresponding deltas.
+
+### Working
+The locking mechanism in the PoolManager contract works as follows:
+
+1. When a user wants to lock, they call the `lock()` function with the data that they want to be passed to the callback.
+2. The `lock()` function pushes the user's address onto the locker queue.
+3. The `lock()` function then calls the `ILockCallback(msg.sender).lockAcquired(data)` callback.
+4. The callback can do whatever it needs to do, such as updating the user's balances or interacting with other contracts.
+5. Once the callback is finished, it returns to the `lock()` function.
+6. The `lock()` function checks if there are any other lockers in the queue. If there are, it pops the next locker off the queue and calls the callback for that locker.
+7. If there are no more lockers in the queue, the `lock()` function returns.
+
+# Lock Data Structure
+
+- **Struct:**
+The `LockData` struct is used to store information related to locks. It has two fields:
+ - `length`: Represents the current number of active locks.
+ - `nonzeroDeltaCount`: Represents the total number of non-zero deltas over all active and completed locks.
+
+- **Library (`LockDataLibrary`):**
+The `LockDataLibrary` is a library used for managing the custom storage implementation of the queue that tracks current lockers.
+
+
+# Example
+Below is the example from a community ["Liquidity Bootstrapping Hook"](https://github.com/kadenzipfel/uni-lbp/blob/main/src/LiquidityBootstrappingHooks.sol) hook that uses the locking mechanism and calls the `modifyPosition` and `swap` functions.
```

Comment:
This example may be too complicated. Maybe use some simple hooks in the v4 repo?